package binary_search.mustdo;

// Problem link - https://leetcode.com/problems/find-peak-element/description/
public class Leetcode_162 {

    public static int findPeakElement(int[] nums) {
        if (nums.length == 1) return 0;

        if (nums[0] > nums[1]) return 0;

        if (nums[nums.length - 1] > nums[nums.length - 2]) return nums.length - 1;

        int start = 1;
        int end = nums.length - 2;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] > nums[mid - 1] && nums[mid] > nums[mid + 1]) {
                return mid;
            } else if (nums[mid] < nums[mid - 1] && nums[mid] < nums[mid + 1]) { //
                // This is a special case to handle  where element present at
                // mid-index smaller than their both the neighbour.
                end = mid - 1;
            } else if (nums[mid] > nums[mid - 1]) {
                start = mid + 1;
            } else if (nums[mid] > nums[mid + 1]) {
                end = mid - 1;
            } else { // case to handle multiple peak.
                start = mid + 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(findPeakElement(new int[]{1, 2, 3, 1}));
        System.out.println(findPeakElement(new int[]{1, 2, 1, 3, 5, 6, 4}));
        System.out.println(findPeakElement(new int[]{1, 2, 1, 2, 1})); // special input use case
        System.out.println(findPeakElement(new int[]{1, 5, 1, 2, 1})); // multiple peak use case
    }
}
